docs/adrs/0009-nimbus-web-sdk-architecture.md

```diff
@@ -0,0 +1,97 @@
+# Nimbus Web SDK Architecture
+
+* Status: proposed
+* Deciders: James Hugman, Charlie Humphreys, Yashika Khurana, Jared Lockhart
+* Date: 2023-15-03
+
+Technical Story: https://mozilla-hub.atlassian.net/browse/EXP-3158
+
+## Context and Problem Statement
+
+This is part of a body of work necessary to support the use of Nimbus within web applications. 
+The current Nimbus SDK is written in such a way that it supports client-oriented experimentation — experiments are downloaded, evaluated, and stored on the client, and a feature store is exposed with the experiment branches applied. 
+In order to support web experimentation, the Nimbus SDK will need to be updated to be stateless, to support a more statically defined set of helper methods, and to have additional support for Python.
+
+Ultimately, the problem we're trying to solve can be boiled down to one question — how can we update the Nimbus SDK to support web applications while continuing to support the existing clients?
+
+## Decision Drivers
+
+* Architecture decisions for Nimbus on the web[<sup>[1.i]</sup>](#links)
+  * The core logic of the existing SDK will be re-used for the Nimbus web service (Cirrus)
+  * The SDK needs to support Python
+  * The SDK needs to be stateless
+
+## Considered Options
+
+* Separate `nimbus-core`, `nimbus-sdk`, and `cirrus-sdk` libraries
+* Cargo features, one library
+* Cargo features, separate library
+* Expose a stateless API alongside the stateful API
+
+## Decision Outcome
+
+[justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force force | … | comes out best (see below)].
+
+### Positive Consequences
+
+* [e.g., improvement of quality attribute satisfaction, follow-up decisions required, …]
+* …
+
+### Negative Consequences
+
+* [e.g., compromising quality attribute, follow-up decisions required, …]
+* …
+
+## Pros and Cons of the Options
+
+### Separate `nimbus-core`, `nimbus-sdk`, and `cirrus-sdk` libraries
+
+For this option, we would take the core Nimbus business logic currently in the `nimbus-sdk` library and move it to a new `nimbus-core` library. 
+The `nimbus-sdk` and `cirrus-sdk` libraries would depend on the `nimbus-core` library, extending the functionality for their various use cases. 
+
+* Good, because separate libraries allow for strong separation of concerns.
+* Bad, because defining and using types becomes more difficult.
+  * Traits will be needed to extend functionality in the Nimbus Core library. While dynamic typing with traits is supported, it is one of the more complex features of Rust.
```

Comment:
I do agree, but I'm also not sure how to go about explaining it succinctly. I've got a code example here: https://github.com/mozilla/application-services/compare/main...jeddai:application-services:create-nimbus-core#diff-02305e8e02a7900352e67be1fd2eef0b5a7c7cf91f4cc3e4559668c124d34e88R11-R19

Reply:
I think linking to an existing example would be good

Reply:
Uh I see, yeah adding the snippet you described here makes a ton of sense!

FWIW, if traits extending other traits is a `Bad` we _might_ be able to work around it. For example, we could:
- Only having one `EventStore` trait and have the consumers (i.e NimbusSDK) deal with whether they are DB backed or not. In practice, this means the NimbusSDK can define a type that implements `EventStore`, and holds on to a DB connection in `self`

- Remove the `Box<dyn..>` and have types be generic over `T: EventStore`. This would mean `NimbusClient` becomes `NimbusClient<T>` and `NimbusClient::event_store`'s type becomes `T`


Regardless - The example clarifies your point a lot and your point still stands, playing around with generics/traits like that is complex.